lib/SILOptimizer/Mandatory/MandatoryInlining.cpp

```diff
@@ -764,137 +764,89 @@ static SILInstruction *tryDevirtualizeApplyHelper(FullApplySite InnerAI,
+                             DominanceAnalysis *DA,
+                             SmallVector<FullApplySite, 8> &collection) {
+  DominanceInfo *DT = DA->get(F);
+  DominanceOrder domOrder(&F->front(), DT, F->size());
+
+  // First collect all the applies
+  SmallVector<FullApplySite, 12> applyInsts;
+  while (SILBasicBlock *block = domOrder.getNext()) {
+    for (auto &inst : *block) {
+      FullApplySite apply = FullApplySite::isa(&inst);
+      if (!apply)
+        continue;
+      applyInsts.push_back(apply);
+    }
   }
 
-  // Add to the current inlining set (immutably, so we only affect the set
-  // during this call and recursive subcalls).
-  CurrentInliningSet = SetFactory.add(CurrentInliningSet, F);
-
-  SmallVector<ParameterConvention, 16> CapturedArgConventions;
-  SmallVector<SILValue, 32> FullArgs;
-  bool needUpdateStackNesting = false;
-
-  // Visiting blocks in reverse order avoids revisiting instructions after block
-  // splitting, which would be quadratic.
-  for (auto BI = F->rbegin(), BE = F->rend(), nextBB = BI; BI != BE;
-       BI = nextBB) {
-    // After inlining, the block iterator will be adjusted to point to the last
-    // block containing inlined instructions. This way, the inlined function
-    // body will be reprocessed within the caller's context without revisiting
-    // any original instructions.
-    nextBB = std::next(BI);
-
-    // While iterating over this block, instructions are inserted and deleted.
-    // To avoid quadratic block splitting, instructions must be processed in
-    // reverse order (block splitting reassigned the parent pointer of all
-    // instructions below the split point).
-    for (auto II = BI->rbegin(); II != BI->rend(); ++II) {
-      FullApplySite InnerAI = FullApplySite::isa(&*II);
-      if (!InnerAI)
-        continue;
+  // Then try to devirtualize them and add them to our collections.
+  for (auto &apply : applyInsts) {
+    // *NOTE* If devirtualization succeeds, devirtInst may not be apply,
+    // but a casted result of apply or even a block argument due to
+    // abstraction changes when calling the witness or class method.
+    auto *devirtInst = tryDevirtualizeApplyHelper(apply, CHA);
+    apply = FullApplySite::isa(devirtInst);
+    if (!apply)
+      continue;
 
-      // *NOTE* If devirtualization succeeds, devirtInst may not be InnerAI,
-      // but a casted result of InnerAI or even a block argument due to
-      // abstraction changes when calling the witness or class method.
-      auto *devirtInst = tryDevirtualizeApplyHelper(InnerAI, CHA);
-      // Restore II to the current apply site.
-      II = devirtInst->getReverseIterator();
-      // If the devirtualized call result is no longer a invalid FullApplySite,
-      // then it has succeeded, but the result is not immediately inlinable.
-      InnerAI = FullApplySite::isa(devirtInst);
-      if (!InnerAI)
-        continue;
+    collection.push_back(apply);
+  }
+
+  return collection.size();
```

Comment:
Why not just return a bool?

Reply:
I was using `.size` as `!.empty`. I'll update it.